enterprise/tailnet/pgcoord_test.go

```diff
@@ -0,0 +1,656 @@
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/slogtest"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/database/dbtestutil"
+	"github.com/coder/coder/enterprise/tailnet"
+	agpl "github.com/coder/coder/tailnet"
+	"github.com/coder/coder/testutil"
+)
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
+}
+
+func TestPGCoordinatorSingle_ClientWithoutAgent(t *testing.T) {
+	t.Parallel()
+	if os.Getenv("DB") == "" {
+		t.Skip("test only with postgres")
+	}
+	store, pubsub := dbtestutil.NewDB(t)
+	ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitSuperLong)
+	defer cancel()
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+	coordinator, err := tailnet.NewPGCoord(ctx, logger, pubsub, store)
+	require.NoError(t, err)
+	defer coordinator.Close()
+
+	agentID := uuid.New()
+	client := newTestClient(t, coordinator, agentID)
+	defer client.close()
+	client.sendNode(&agpl.Node{PreferredDERP: 10})
+	require.Eventually(t, func() bool {
+		clients, err := store.GetTailnetClientsForAgent(ctx, agentID)
+		if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+			t.Fatalf("database error: %v", err)
+		}
+		if len(clients) == 0 {
+			return false
+		}
+		var node agpl.Node
+		err = json.Unmarshal(clients[0].Node, &node)
+		assert.NoError(t, err)
+		assert.Equal(t, 10, node.PreferredDERP)
+		return true
+	}, testutil.WaitShort, testutil.IntervalFast)
+
+	err = client.close()
+	require.NoError(t, err)
+	<-client.errChan
+	<-client.closeChan
+	assertEventuallyNoClientsForAgent(ctx, t, store, agentID)
+}
+
+func TestPGCoordinatorSingle_AgentWithoutClients(t *testing.T) {
+	t.Parallel()
+	if os.Getenv("DB") == "" {
+		t.Skip("test only with postgres")
+	}
+	store, pubsub := dbtestutil.NewDB(t)
+	ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitSuperLong)
+	defer cancel()
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+	coordinator, err := tailnet.NewPGCoord(ctx, logger, pubsub, store)
+	require.NoError(t, err)
+	defer coordinator.Close()
+
+	agent := newTestAgent(t, coordinator)
+	defer agent.close()
+	agent.sendNode(&agpl.Node{PreferredDERP: 10})
+	require.Eventually(t, func() bool {
+		agents, err := store.GetTailnetAgents(ctx, agent.id)
+		if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+			t.Fatalf("database error: %v", err)
+		}
+		if len(agents) == 0 {
+			return false
+		}
+		var node agpl.Node
+		err = json.Unmarshal(agents[0].Node, &node)
+		assert.NoError(t, err)
+		assert.Equal(t, 10, node.PreferredDERP)
+		return true
+	}, testutil.WaitShort, testutil.IntervalFast)
+	err = agent.close()
+	require.NoError(t, err)
+	<-agent.errChan
+	<-agent.closeChan
+	assertEventuallyNoAgents(ctx, t, store, agent.id)
+}
+
+func TestPGCoordinatorSingle_AgentWithClient(t *testing.T) {
+	t.Parallel()
+	if os.Getenv("DB") == "" {
+		t.Skip("test only with postgres")
+	}
```

Comment:
nit: actually it could be extracted to a helper function if we decide to replace `DB` with something else.